### PR TITLE
[REBASE only]: Migrate notebooks and add new amalthea

### PR DESCRIPTION
**Pull request by olevski**
This combines all changes for the notebooks migration and new amalthea operator.

**NOTE: This should be rebased on main. DO NOT SQUASH AND MERGE. WHEN updating this branch remember to REABASE only too.**

/deploy renku=feat-jupyter-free-sessions amalthea-sessions=main renku-ui=andrea/jupyter-free-build renku-gateway=master extra-values=amalthea-sessions.deployCrd=false

### CODE REVIEW

**Comment by olevski** [comment]
This change is part of the following stack:

- #418 ◀
    - #338
    - #375
        - #387
            - #411
    - #434
        - #439
            - #440

<sub>Change managed by [git-spice](https://abhinav.github.io/git-spice/).</sub>

**Review by leafty — COMMENTED** [review]
Cannot read logs when some pods are not initialized:

```
Srv 0 2024-10-30 14:31:38 +0000 ERROR: Exception occurred while handling uri: 'https://renku-ci-ds-478.dev.renku.ch/api/data/sessions/flora-thieba-db93d83e77cc/logs?max_lines=250'
Traceback (most recent call last):
  File "/app/env/lib/python3.12/site-packages/kr8s/_api.py", line 166, in call_api
    response.raise_for_status()
  File "/app/env/lib/python3.12/site-packages/httpx/_models.py", line 763, in raise_for_status
    raise HTTPStatusError(message, request=request, response=self)
httpx.HTTPStatusError: Client error '400 Bad Request' for url 'https://10.43.0.1/api/v1/namespaces/renku-ci-ds-478/pods/flora-thieba-db93d83e77cc-0/log?container=amalthea-session&tailLines=250'
For more information check: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "handle_request", line 102, in handle_request
    if TYPE_CHECKING:
                      
  File "/app/env/lib/python3.12/site-packages/renku_data_services/base_api/auth.py", line 36, in decorated_function
    response = await f(request, user, *args, **kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/env/lib/python3.12/site-packages/sanic_ext/extras/validation/decorator.py", line 75, in decorated_function
    retval = await retval
             ^^^^^^^^^^^^
  File "/app/env/lib/python3.12/site-packages/renku_data_services/notebooks/blueprints.py", line 651, in _handler
    logs = await self.nb_config.k8s_v2_client.get_server_logs(session_id, user.id, query.max_lines)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/env/lib/python3.12/site-packages/renku_data_services/notebooks/api/classes/k8s_client.py", line 454, in get_server_logs
    return await self.renku_ns_client.get_pod_logs(pod_name, max_log_lines)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/env/lib/python3.12/site-packages/renku_data_services/notebooks/api/classes/k8s_client.py", line 85, in get_pod_logs
    clogs: list[str] = [clog async for clog in pod.logs(container=container, tail_lines=max_log_lines)]
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/env/lib/python3.12/site-packages/kr8s/_objects.py", line 890, in logs
    async with self.api.call_api(
               ^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/contextlib.py", line 210, in __aenter__
    return await anext(self.gen)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/app/env/lib/python3.12/site-packages/kr8s/_api.py", line 185, in call_api
    error = e.response.json()
            ^^^^^^^^^^^^^^^^^
  File "/app/env/lib/python3.12/site-packages/httpx/_models.py", line 766, in json
    return jsonlib.loads(self.content, **kwargs)
                         ^^^^^^^^^^^^
  File "/app/env/lib/python3.12/site-packages/httpx/_models.py", line 572, in content
    raise ResponseNotRead()
httpx.ResponseNotRead: Attempted to access streaming response content, without having called `read()`.
```

**Comment by olevski** [comment]
@leafty this PR fixes the issues you found: https://github.com/SwissDataScienceCenter/renku-data-services/pull/504

**Comment by leafty** [comment]
There is still an error while fetching logs:

```
Srv 0 2024-11-07 08:14:54 +0000 ERROR: Exception occurred while handling uri: 'https://renku-ci-ui-3341.dev.renku.ch/api/data/sessions/flora-thieba-8f2adbaccd0d/logs?max_lines=250'
Traceback (most recent call last):
  File "/app/env/lib/python3.12/site-packages/kr8s/_api.py", line 166, in call_api
    response.raise_for_status()
  File "/app/env/lib/python3.12/site-packages/httpx/_models.py", line 763, in raise_for_status
    raise HTTPStatusError(message, request=request, response=self)
httpx.HTTPStatusError: Client error '400 Bad Request' for url 'https://10.43.0.1/api/v1/namespaces/renku-ci-ui-3341/pods/flora-thieba-8f2adbaccd0d-0/log?container=amalthea-session&tailLines=250'
For more information check: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "handle_request", line 102, in handle_request
    if TYPE_CHECKING:
                      
  File "/app/env/lib/python3.12/site-packages/renku_data_services/base_api/auth.py", line 36, in decorated_function
    response = await f(request, user, *args, **kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/env/lib/python3.12/site-packages/sanic_ext/extras/validation/decorator.py", line 75, in decorated_function
    retval = await retval
             ^^^^^^^^^^^^
  File "/app/env/lib/python3.12/site-packages/renku_data_services/notebooks/blueprints.py", line 650, in _handler
    logs = await self.nb_config.k8s_v2_client.get_server_logs(session_id, user.id, query.max_lines)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/env/lib/python3.12/site-packages/renku_data_services/notebooks/api/classes/k8s_client.py", line 454, in get_server_logs
    return await self.renku_ns_client.get_pod_logs(pod_name, max_log_lines)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/env/lib/python3.12/site-packages/renku_data_services/notebooks/api/classes/k8s_client.py", line 85, in get_pod_logs
    clogs: list[str] = [clog async for clog in pod.logs(container=container, tail_lines=max_log_lines)]
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/env/lib/python3.12/site-packages/kr8s/_objects.py", line 890, in logs
    async with self.api.call_api(
               ^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/contextlib.py", line 210, in __aenter__
    return await anext(self.gen)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/app/env/lib/python3.12/site-packages/kr8s/_api.py", line 185, in call_api
    error = e.response.json()
            ^^^^^^^^^^^^^^^^^
  File "/app/env/lib/python3.12/site-packages/httpx/_models.py", line 766, in json
    return jsonlib.loads(self.content, **kwargs)
                         ^^^^^^^^^^^^
  File "/app/env/lib/python3.12/site-packages/httpx/_models.py", line 572, in content
    raise ResponseNotRead()
httpx.ResponseNotRead: Attempted to access streaming response content, without having called `read()`.
```

**Comment by leafty** [comment]
The data-service at commit `fe07dab` does not correctly report the status of running sessions:

```json
[
    {
        "image": "renku/renkulab-py:3.10-0.23.0",
        "name": "flora-thieba-8f2adbaccd0d",
        "resources": {
            "requests": {
                "cpu": 0.1,
                "memory": 0,
                "storage": 1
            }
        },
        "started": "2024-11-07T08:14:41Z",
        "status": {
            "state": "starting",
            "will_hibernate_at": "2024-11-07T08:14:41Z",
            "ready_containers": 0,
            "total_containers": 0
        },
        "url": "https://renku-ci-ui-3341.dev.renku.ch/sessions/flora-thieba-8f2adbaccd0d/",
        "project_id": "01JC0GF2QNHPKZXM9978T9JNJV",
        "launcher_id": "01JC2TZSZCSFYQ8XCVMETHPAPA",
        "resource_class_id": 3
    }
]
```

But from kubernetes (session is fully running):

```
flora-thieba-8f2adbaccd0d-0   3/3     Running   0          3m43s   10.42.12.4   worker-dev.renku-user-sessions-3   <none>           <none>
```

**Review by leafty — COMMENTED** [review]
#504 has been swallowed by the void it seems. Internal gitlab repositories are broken.

**Comment by olevski** [comment]
> https://github.com/SwissDataScienceCenter/renku-data-services/pull/504 has been swallowed by the void it seems. Internal gitlab repositories are broken.

Yeah I probably messed something up in one of the many rebases... Here are the same commits from #504 in a new PR: #518

**Comment by olevski** [comment]
> The data-service at commit fe07dab does not correctly report the status of running sessions:
@leafty this does not occur anymore.

**Review by leafty — DISMISSED** [review]
The sessions seem to be working correctly now 👏 🎉

**Review by leafty — COMMENTED** [review]
Note: when resuming a paused session, the first reported statuses are `N/N containers ready`.

```
ready_containers: 6
state: "starting"
total_containers: 6
```

After a bit of time the status is properly reported and goes backward to `0/N`.

File: components/renku_data_services/migrations/versions/1ef98b967767_add_command_and_args_to_environment.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+down_revision = "584598f3b769"
+branch_labels = None
+depends_on = None
+
+default_command = ["sh", "-c"]
+default_args = [
+    "jupyter server --ServerApp.ip=0.0.0.0 --ServerApp.port=8888 --ServerApp.base_url=$RENKU_BASE_URL_PATH "
+    '--ServerApp.token="" --ServerApp.password="" --ServerApp.allow_remote_access=true '
+    "--ContentsManager.allow_hidden=true --ServerApp.allow_origin=*",
+]
```
Comment (leafty) [inline components/renku_data_services/migrations/versions/1ef98b967767_add_command_and_args_to_environment.py:24]: See: https://renku-ci-ds-517.dev.renku.ch/v2/projects/flora.thiebaut/flora-test.

This runs the `/post-init.sh` script if present.

```suggestion
default_args = [
    "/entrypoint.sh  jupyter server --ServerApp.ip=0.0.0.0 --ServerApp.port=8888 --ServerApp.base_url=$RENKU_BASE_URL_PATH "
    '--ServerApp.token="" --ServerApp.password="" --ServerApp.allow_remote_access=true '
    "--ContentsManager.allow_hidden=true --ServerApp.allow_origin=*",
]
```

**Comment by olevski** [comment]
>Note: when resuming a paused session, the first reported statuses are N/N containers ready.
After a bit of time the status is properly reported and goes backward to 0/N.

This should be fixed in amalthea. And I will do that.

**Review by leafty — APPROVED** [review]

**Comment by leafty** [comment]
There is a test failing because of this -> https://github.com/SwissDataScienceCenter/renku-data-services/issues/525.

**Comment by olevski** [comment]
> There is a test failing because of this -> https://github.com/SwissDataScienceCenter/renku-data-services/issues/525.

I saw that I just used a tagged image in the test which passes the tests for now. I dont want to mess with the image checks further here.

**Pull request merged**